File: gtci/00-warmup/cpp/1-is-pangram.cpp
#include <cctype>
#include <iostream>
#include <unordered_set>

using namespace std;

class Solution {
   public:
    bool checkIfPangram(string sentence) {
        unordered_set<char> letters{};

        for (char ch : sentence) {
            if (!isalpha(ch)) continue;

            if (letters.find(tolower(ch)) ==
                letters.end())  // contains is a > C++20 feature
                letters.insert(tolower(ch));
        }

        return letters.size() == 26;
    }
};

#include <cassert>

int main() {
    Solution sol;
    assert(sol.checkIfPangram("The quick brown fox jumps over the lazy dog") ==
           true);
    assert(sol.checkIfPangram("Hello, World!") == false);
    assert(sol.checkIfPangram("abcdefghijklmnopqrstuvwxyz") == true);
    assert(sol.checkIfPangram("You shall not pass!") == false);
    assert(sol.checkIfPangram("Pack my box with five dozen liquor jugs.") ==
           true);
    std::cout << "All test cases passed." << std::endl;
    return 0;
}

File: gtci/00-warmup/python3/1-is-pangram.py
def checkIfPangram(sentence):
    letters = {}

    for letter in sentence:
        if not letter.isalpha():
            continue

        letter_lower = letter.lower()

        if letter_lower not in letters:
            letters[letter_lower] = True

    return len(letters) == 26


if __name__ == "__main__":
    assert checkIfPangram("The quick brown fox jumps over the lazy dog") == True
    assert checkIfPangram("You shall not pass!") == False
    assert checkIfPangram("Pack my box with five dozen liquor jugs.") == True
    print("All test cases passed.")
